pkg/build/strategies/dockerfile/dockerfile.go

```diff
@@ -0,0 +1,323 @@
+		api.UserScripts,
+	}
+)
+
+// Dockerfile builders produce a Dockerfile rather than an image.
+// Building the dockerfile w/ the right context will result in
+// an application image being produced.
+type Dockerfile struct {
+	config           *api.Config
+	fs               fs.FileSystem
+	uploadScriptsDir string
+	uploadSrcDir     string
+	sourceInfo       *git.SourceInfo
+	result           *api.Result
+	ignorer          build.Ignorer
+}
+
+// New creates a Dockerfile builder.
+func New(config *api.Config, fs fs.FileSystem) (*Dockerfile, error) {
+	/*
+		excludePattern, err := regexp.Compile(config.ExcludeRegExp)
+		if err != nil {
+			return nil, err
+		}
+	*/
+
+	return &Dockerfile{
+		config: config,
+		fs:     fs,
+		// where we will get the assemble/run scripts from on the host machine,
+		// if any are provided.
+		uploadScriptsDir: api.UploadScripts,
+		uploadSrcDir:     api.Source,
+		result:           &api.Result{},
+		ignorer:          &ignore.DockerIgnorer{},
+	}, nil
+}
+
+// Build produces a Dockerfile that when run with the correct filesystem
+// context, will produce the application image.
+func (builder *Dockerfile) Build(config *api.Config) (*api.Result, error) {
+	if config.BuilderImage == "" {
+		builder.result.BuildInfo.FailureReason = utilstatus.NewFailureReason(
+			utilstatus.ReasonGenericS2IBuildFailed,
+			utilstatus.ReasonMessageGenericS2iBuildFailed,
+		)
+		return builder.result, errors.New("builder image name cannot be empty")
+	}
+
+	if err := builder.Prepare(config); err != nil {
+		return builder.result, err
+	}
+
+	if err := builder.CreateDockerfile(config); err != nil {
+		builder.result.BuildInfo.FailureReason = utilstatus.NewFailureReason(
+			utilstatus.ReasonDockerfileCreateFailed,
+			utilstatus.ReasonMessageDockerfileCreateFailed,
+		)
+		return builder.result, err
+	}
+
+	builder.result.Success = true
+
+	return builder.result, nil
+}
+
+// CreateDockerfile takes the various inputs and creates the Dockerfile used by
+// the docker cmd to create the image produced by s2i.
+func (builder *Dockerfile) CreateDockerfile(config *api.Config) error {
+	glog.V(4).Infof("Constructing image build context directory at %s", builder.config.WorkingDir)
+	buffer := bytes.Buffer{}
+
+	if len(config.ImageWorkDir) == 0 {
+		config.ImageWorkDir = "/opt/app-root/src"
+	}
+
+	imageUser := config.AssembleUser
+
+	// where files will land inside the new image.
+	// TODO: get this from the image metadata
+	scriptsDestDir := filepath.Join(getDestination(config), "scripts")
+	sourcesDestDir := filepath.Join(getDestination(config), "src")
+
+	buffer.WriteString(fmt.Sprintf("FROM %s\n", builder.config.BuilderImage))
+	buffer.WriteString(fmt.Sprintf("USER %s\n", imageUser))
+
+	generatedLabels := util.GenerateOutputImageLabels(builder.sourceInfo, builder.config)
+	if len(generatedLabels) > 0 {
+		buffer.WriteString("LABEL")
+		for k, v := range generatedLabels {
+			buffer.WriteString(fmt.Sprintf(" %s=%s", k, v))
```

Comment:
we care about it less than we care about creating unnecessary layers...every line in a dockerfile is an additional layer, so that's why i did it this way.

Reply:
I was thinking of using `\` to keep it all in one layer:
```
LABEL <key>=<value> \
            <key2>=<value2> \
           ...
```

Reply:
i can take a shot at that.